independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RemovedBeanImpl.java

```diff
@@ -4,21 +4,30 @@
 import io.quarkus.arc.RemovedBean;
 import java.lang.annotation.Annotation;
 import java.lang.reflect.Type;
+import java.util.Arrays;
+import java.util.HashSet;
 import java.util.Set;
 
 public final class RemovedBeanImpl implements RemovedBean {
 
+    /**
+     * Implementation note: this class needs to be optimised to
+     * minimize the size of retained memory: runtime efficiency is less important.
+     */
+    private static final Type[] EMPTY_TYPES = new Type[] {};
+    private static final Annotation[] EMPTY_QUALIFIERS = new Annotation[] {};
+    private static final Annotation[] DEFAULT_QUALIFIERS = Qualifiers.DEFAULT_QUALIFIERS.toArray(EMPTY_QUALIFIERS);
+
     private final Kind kind;
     private final String description;
-    private final Set<Type> types;
-    private final Set<Annotation> qualifiers;
+    private final Type[] types;
+    private final Annotation[] qualifiers;
```

Comment:
I think that most of the beans use the default qualifiers (aka the shared set), although many of the removed beans in any quickstart will be `[@Any, @ConfigProperty]` (because SR Config registers a dozen of config producer methods). 

I believe that we could modify the `io.quarkus.arc.processor.ComponentsProviderGenerator`  to share all sets of qualifiers with no runtime penalty.

Reply:
sharing sets would be nice, but is it worth it for the removed beans? By using an array we already reduce the retained memory quite a bit, not sure if it's worth spending more time in trying to safe some KBs as this is all just references overhead.

If you have such ideas, would be nice to apply the optimisation to the non-removed beans? I focused on the removed ones as I felt it was odd to spend significant memory just to provide that error message (and because it was simpler for me to get a full picture of the use cases), but there's certainly more to be won in the area of the actually active beans.

Reply:
> If you have such ideas, would be nice to apply the optimisation to the non-removed beans?

+1